go/pkg/cas/upload.go

```diff
@@ -38,6 +39,20 @@ type UploadInput struct {
 	// Ignored if Path is not empty.
 	Content []byte
 
+	// PathInclude is a file/dir filter. If PathInclude is not nil and the absolute
+	// path of a file/dir does not match this regexp, then the file/dir is skipped.
+	// If the Path is a directory, then the filter is evaluated against each file
+	// in the subtree.
+	// See ErrSkip comments for more details on semantics.
+	//
+	// This field has no effect if Path is empty.
+	PathInclude *regexp.Regexp
```

Comment:
I think we don't use this?

Reply:
we don't, but having PathExclude without PathInclude seems asymmetric. 
Since this is a general-purpose lib, I figured why not implement both right away.

Reply:
I see, but my preference is to follow https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it
Resolving asymmetric feeling is not important compared to not having unnecessary code and tests for that?

Reply:
@ramymedhat to decide

Reply:
The command proto also only defines input [exclusions](https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/api/command/command.proto#L70), so I think just having exclusions is sufficient for now.

Reply:
Done.

Reply:
Actually, we could use `PathInclude` to filter files under the input root for reclient, instead of the callback - and that gets saved on the cache key.

Wdyt?

Reply:
- reclient has one execRoot for the entire invocation, so not sure it would help reclient
- it would have a different semantics from now. Currently it returns an error https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/pkg/client/tree.go#L96, hence the Callbacks func signature that let's you return an error

Reply:
I'm hoping to avoid callbacks since they don't get cached, but I agree that error-ing at a source outside the root is essential for reclient. But the best option I can think of is naming a `InputRoot` (instead of `PathInclude`, as to not confuse users for the different semantics) that is just a prefix + error-ing.

This is by far less powerful than a callback, but I _think_ from an RE perspective that's good enough, since the concept of an input/exec root is quite central to the API. Wdyt?

I don't think the performance impact would be substantial (besides eg potentially caching errors, and avoiding one callstack that the go compiler might as well inline) to block on this though.